### PR TITLE
Hashable -> Hasher; SimpleMap upgrade; No "SimpleHashFromBinary"

### PR DESCRIPTION
**Pull request by jaekwon**
Update SimpleMap to hash both keys and values for benefit:
* You can blind the key even with a Merkle proof.
* It's potentially shorter than a long key
* Always hash the value, drop "SimpleHashFromBinary(ies)"
Hashable is Hasher
Don't assume go-wire

### CODE REVIEW

File: merkle/simple_map.go [hunk 1]
```diff
@@ -18,25 +18,25 @@ func NewSimpleMap() *SimpleMap {
...
 
-	// Is value Hashable?
-	var vBytes []byte
-	if hashable, ok := value.(Hashable); ok {
-		vBytes = hashable.Hash()
-	} else {
-		vBytes, _ = wire.MarshalBinary(value)
-	}
+	// Hash the key to blind it... why not?
+	khash := SimpleHashFromBytes([]byte(key))
```
Comment (melekes) [inline merkle/simple_map.go]: why not declare `key` as `Hasher` too and call `key.Hash()` here?

File: merkle/simple_tree.go [hunk 2]
```diff
@@ -54,28 +54,25 @@ func SimpleHashFromHashes(hashes [][]byte) []byte {
 	}
 }
 
-// Convenience for SimpleHashFromHashes.
-func SimpleHashFromBinaries(items []interface{}) []byte {
-	hashes := make([][]byte, len(items))
-	for i, item := range items {
-		hashes[i] = SimpleHashFromBinary(item)
+// NOTE: Do not implement this, use SimpleHashFromByteslices instead.
```
Comment (melekes) [inline merkle/simple_tree.go]: Do not implement what?

File: merkle/types.go [hunk 3]
```diff
@@ -18,6 +18,10 @@ type Tree interface {
 	IterateRange(start []byte, end []byte, ascending bool, fx func(key []byte, value []byte) (stop bool)) (stopped bool)
 }
 
-type Hashable interface {
+type Hasher interface {
 	Hash() []byte
 }
+
+type Byteser interface {
```
Comment (melekes) [inline merkle/types.go]: Do we need to define this interface?

File: merkle/simple_map.go [hunk 4]
```diff
@@ -18,25 +18,25 @@ func NewSimpleMap() *SimpleMap {
...
-		Value: vBytes,
+		Key:   khash,
+		Value: vhash,
 	})
 }
 
-// Merkle root hash of items sorted by key.
-// NOTE: Behavior is undefined when key is duplicate.
+// Merkle root hash of items sorted by key
+// (UNSTABLE: and by value too if duplicate key).
```
Comment (ebuchman) [inline merkle/simple_map.go]: UNSTABLE ?

File: merkle/simple_map.go (hunk 1 above)
Comment (jaekwon) [reply merkle/simple_map.go]: I think most cases will just need a string or []byte, so optimizing for the typical case.

File: merkle/simple_map.go (hunk 4 above)
Comment (jaekwon) [reply merkle/simple_map.go]: behavior may change.

File: merkle/simple_tree.go (hunk 2 above)
Comment (jaekwon) [reply merkle/simple_tree.go]: SimpleHashFromBytesers.  Better to just use SimpleHashFromByteslices.

File: merkle/types.go (hunk 3 above)
Comment (jaekwon) [reply merkle/types.go]: No.

**Pull request merged**